enigma/src/main/java/cuchaz/enigma/utils/ZippingIterator.java

```diff
@@ -0,0 +1,16 @@
+package cuchaz.enigma.utils;
+
+import java.util.Iterator;
+import java.util.function.BiFunction;
+
+public record ZippingIterator<T1, T2, R>(Iterator<T1> left, Iterator<T2> right, BiFunction<T1, T2, R> join) implements Iterator<R> {
```

Comment:
As stated before the only usage of this class could be replaced with a relatively simple loop, so there should be no need for this class after that.